File: http4k-multipart/src/test/java/org/http4k/multipart/StreamingMultipartFormHappyTests.java
package org.http4k.multipart;

import kotlin.Pair;
import org.http4k.multipart.exceptions.AlreadyClosedException;
import org.http4k.multipart.part.StreamingPart;
import org.junit.Test;

import java.io.ByteArrayInputStream;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.util.Iterator;
import java.util.Map;

import static java.util.Arrays.asList;
import static org.hamcrest.CoreMatchers.equalTo;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.core.IsNot.not;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

public class StreamingMultipartFormHappyTests {

    public static final String CR_LF = "\r\n";

    @Test
    public void uploadEmptyContents() throws Exception {
        String boundary = "-----1234";
        Iterator<StreamingPart> form = getMultipartFormParts(boundary, new ValidMultipartFormBuilder(boundary).build());

        assertThereAreNoMoreParts(form);
    }

    @Test
    public void uploadEmptyFile() throws Exception {
        String boundary = "-----2345";
        Iterator<StreamingPart> form = getMultipartFormParts(boundary, new ValidMultipartFormBuilder(boundary)
                .file("aFile", "", "doesnt/matter", "").build());

        assertFilePart(form, "aFile", "", "doesnt/matter", "");

        assertThereAreNoMoreParts(form);
    }

    @Test
    public void hasNextIsIdempotent() throws Exception {
        String boundary = "-----2345";
        Iterator<StreamingPart> form = getMultipartFormParts(boundary, new ValidMultipartFormBuilder(boundary)
                .file("aFile", "", "application/octet-stream", "")
                .file("anotherFile", "", "application/octet-stream", "").build());

        assertThereAreMoreParts(form);
        assertThereAreMoreParts(form);

        form.next();

        assertThereAreMoreParts(form);
        assertThereAreMoreParts(form);

        form.next();

        assertThereAreNoMoreParts(form);
        assertThereAreNoMoreParts(form);
    }

    @Test
    public void uploadEmptyField() throws Exception {
        String boundary = "-----3456";
        Iterator<StreamingPart> form = getMultipartFormParts(boundary, new ValidMultipartFormBuilder(boundary)
                .field("aField", "").build());

        assertFieldPart(form, "aField", "");

        assertThereAreNoMoreParts(form);
    }

    @Test
    public void uploadSmallFile() throws Exception {
        String boundary = "-----2345";
        Iterator<StreamingPart> form = getMultipartFormParts(boundary, new ValidMultipartFormBuilder(boundary)
                .file("aFile", "file.name", "application/octet-stream", "File contents here").build());

        assertFilePart(form, "aFile", "file.name", "application/octet-stream", "File contents here");

        assertThereAreNoMoreParts(form);
    }

    @Test
    public void uploadSmallFileAsAttachment() throws Exception {
        String boundary = "-----4567";
        Iterator<StreamingPart> form = getMultipartFormParts(boundary, new ValidMultipartFormBuilder(boundary)
                .file("beforeFile", "before.txt", "application/json", "[]")
                .startMultipart("multipartFieldName", "7890")
                .attachment("during.txt", "plain/text", "Attachment contents here")
                .attachment("during2.txt", "plain/text", "More text here")
                .endMultipart()
                .file("afterFile", "after.txt", "application/json", "[]")
                .build());

        assertFilePart(form, "beforeFile", "before.txt", "application/json", "[]");
        assertFilePart(form, "multipartFieldName", "during.txt", "plain/text", "Attachment contents here");
        assertFilePart(form, "multipartFieldName", "during2.txt", "plain/text", "More text here");
        assertFilePart(form, "afterFile", "after.txt", "application/json", "[]");

        assertThereAreNoMoreParts(form);
    }


    @Test
    public void uploadSmallField() throws Exception {
        String boundary = "-----3456";
        Iterator<StreamingPart> form = getMultipartFormParts(boundary, new ValidMultipartFormBuilder(boundary)
                .field("aField", "Here is the value of the field\n").build());

        assertFieldPart(form, "aField", "Here is the value of the field\n");

        assertThereAreNoMoreParts(form);
    }

    @Test
    public void uploadMultipleFilesAndFields() throws Exception {
        String boundary = "-----1234";
        Iterator<StreamingPart> form = getMultipartFormParts(boundary,
                new ValidMultipartFormBuilder(boundary)
                        .file("file", "foo.tab", "text/whatever", "This is the content of the file\n")
                        .field("field", "fieldValue" + CR_LF + "with cr lf")
                        .field("multi", "value1")
                        .file("anotherFile", "BAR.tab", "text/something", "This is another file\n")
                        .field("multi", "value2")
                        .build());

        assertFilePart(form, "file", "foo.tab", "text/whatever", "This is the content of the file\n");
        assertFieldPart(form, "field", "fieldValue" + CR_LF + "with cr lf");
        assertFieldPart(form, "multi", "value1");
        assertFilePart(form, "anotherFile", "BAR.tab", "text/something", "This is another file\n");
        assertFieldPart(form, "multi", "value2");

        assertThereAreNoMoreParts(form);
    }

    @Test
    public void uploadFieldsWithMultilineHeaders() throws Exception {
        String boundary = "-----1234";
        Iterator<StreamingPart> form = getMultipartFormParts(boundary,
                new ValidMultipartFormBuilder(boundary)
                        .rawPart(
                                "Content-Disposition: form-data; \r\n" +
                                        "\tname=\"field\"\r\n" +
                                        "\r\n" +
                                        "fieldValue")
                        .rawPart(
                                "Content-Disposition: form-data;\r\n" +
                                        "     name=\"multi\"\r\n" +
                                        "\r\n" +
                                        "value1")
                        .field("multi", "value2")
                        .build());

        assertFieldPart(form, "field", "fieldValue");
        assertFieldPart(form, "multi", "value1");
        assertFieldPart(form, "multi", "value2");

        assertThereAreNoMoreParts(form);
    }

    @Test
    public void partsCanHaveLotsOfHeaders() throws Exception {
        String boundary = "-----1234";
        Iterator<StreamingPart> form = getMultipartFormParts(boundary,
                new ValidMultipartFormBuilder(boundary)
                        .part("This is the content of the file\n",
                                new Pair("Content-Disposition", asList(new Pair("form-data", null), new Pair("name", "fileFieldName"), new Pair("filename", "filename.txt"))),
                                new Pair("Content-Type", asList(new Pair("plain/text", null))),
                                new Pair("Some-header", asList(new Pair("some value", null))))
                        .part("This is the content of the field\n",
                                new Pair("Content-Disposition", asList(new Pair("form-data", null), new Pair("name", "fieldFieldName"))),
                                new Pair("Another-header", asList(new Pair("some-key", "some-value")))
                        )
                        .build());

        StreamingPart file = assertFilePart(form, "fileFieldName", "filename.txt", "plain/text", "This is the content of the file\n");

        Map<String, String> fileHeaders = file.getHeaders();
        assertThat(fileHeaders.size(), equalTo(3));
        assertThat(fileHeaders.get("Content-Disposition"), equalTo("form-data; name=\"fileFieldName\"; filename=\"filename.txt\""));
        assertThat(fileHeaders.get("Content-Type"), equalTo("plain/text"));
        assertThat(fileHeaders.get("Some-header"), equalTo("some value"));

        StreamingPart field = assertFieldPart(form, "fieldFieldName", "This is the content of the field\n");

        Map<String, String> fieldHeaders = field.getHeaders();
        assertThat(fieldHeaders.size(), equalTo(2));
        assertThat(fieldHeaders.get("Content-Disposition"), equalTo("form-data; name=\"fieldFieldName\""));
        assertThat(fieldHeaders.get("Another-header"), equalTo("some-key=\"some-value\""));

        assertThereAreNoMoreParts(form);
    }

    @Test
    public void closedPartsCannotBeReadFrom() throws Exception {
        String boundary = "-----2345";
        Iterator<StreamingPart> form = getMultipartFormParts(boundary, new ValidMultipartFormBuilder(boundary)
                .file("aFile", "file.name", "application/octet-stream", "File contents here").build());

        StreamingPart file = form.next();

        //noinspection StatementWithEmptyBody
        while (file.inputStream.read() > 0) {
            // keep reading.
        }

        assertThat(file.inputStream.read(), equalTo(-1));
        file.inputStream.close();
        file.inputStream.close(); // can close multiple times
        try {
            int ignored = file.inputStream.read();
            fail("Should have complained that the StreamingPart has been closed " + ignored);
        } catch (AlreadyClosedException e) {
            // pass
        }
    }

    @Test
    public void readingPartsContentsAsStringClosesStream() throws Exception {
        String boundary = "-----2345";
        Iterator<StreamingPart> form = getMultipartFormParts(boundary, new ValidMultipartFormBuilder(boundary)
                .file("aFile", "file.name", "application/octet-stream", "File contents here").build());

        StreamingPart file = form.next();
        file.getContentsAsString();

        try {
            int ignored = file.inputStream.read();
            fail("Should have complained that the StreamingPart has been closed " + ignored);
        } catch (AlreadyClosedException e) {
            // pass
        }

        file.inputStream.close(); // can close multiple times
    }

    @Test
    public void gettingNextPartClosesOldPart() throws Exception {
        String boundary = "-----2345";
        Iterator<StreamingPart> form = getMultipartFormParts(boundary, new ValidMultipartFormBuilder(boundary)
                .file("aFile", "file.name", "application/octet-stream", "File contents here")
                .file("anotherFile", "your.name", "application/octet-stream", "Different file contents here").build());

        StreamingPart file1 = form.next();

        StreamingPart file2 = form.next();

        assertThat(file1, not(equalTo(file2)));

        try {
            int ignored = file1.inputStream.read();
            fail("Should have complained that the StreamingPart has been closed " + ignored);
        } catch (AlreadyClosedException e) {
            // pass
        }

        file1.inputStream.close(); // can close multiple times

        assertThat(file2.getContentsAsString(), equalTo("Different file contents here"));
    }

    @Test
    public void canLoadComplexRealLifeSafariExample() throws Exception {
        Iterator<StreamingPart> parts = StreamingMultipartFormParts.Companion.parse(
                "----WebKitFormBoundary6LmirFeqsyCQRtbj".getBytes(StandardCharsets.UTF_8),
                new FileInputStream("examples/safari-example.multipart"),
                StandardCharsets.UTF_8
        ).iterator();

        assertFieldPart(parts, "articleType", "obituary");

        assertRealLifeFile(parts, "simple7bit.txt", "text/plain");
        assertRealLifeFile(parts, "starbucks.jpeg", "image/jpeg");
        assertRealLifeFile(parts, "utf8\uD83D\uDCA9.file", "application/octet-stream");
        assertRealLifeFile(parts, "utf8\uD83D\uDCA9.txt", "text/plain");

    }

    @Test
    public void canLoadComplexRealLifeChromeExample() throws Exception {
        Iterator<StreamingPart> parts = StreamingMultipartFormParts.Companion.parse(
                "----WebKitFormBoundaryft3FGhOMTYoOkCCc".getBytes(StandardCharsets.UTF_8),
                new FileInputStream("examples/chrome-example.multipart"),
                StandardCharsets.UTF_8
        ).iterator();

        assertFieldPart(parts, "articleType", "obituary");

        assertRealLifeFile(parts, "simple7bit.txt", "text/plain");
        assertRealLifeFile(parts, "starbucks.jpeg", "image/jpeg");
        assertRealLifeFile(parts, "utf8\uD83D\uDCA9.file", "application/octet-stream");
        assertRealLifeFile(parts, "utf8\uD83D\uDCA9.txt", "text/plain");

    }

    private void assertRealLifeFile(Iterator<StreamingPart> parts, String fileName, String contentType) throws IOException {
        StreamingPart file = parts.next();
        assertThat("field name", file.getFieldName(), equalTo("uploadManuscript"));
        assertThat("file name", file.getFileName(), equalTo(fileName));
        assertThat("content type", file.getContentType(), equalTo(contentType));
        assertPartIsNotField(file);
        compareStreamToFile(file);
    }

    public static void compareStreamToFile(StreamingPart file) throws IOException {
        InputStream formFile = file.inputStream;
        compareStreamToFile(formFile, file.getFileName());
    }

    public static void compareStreamToFile(InputStream actualSream, String fileName) throws IOException {
        InputStream original = new FileInputStream("examples/" + fileName);
        compareOneStreamToAnother(actualSream, original);
    }

    public static void compareOneStreamToAnother(InputStream actualStream, InputStream expectedStream) throws IOException {
        int index = 0;
        while (true) {
            int actual = actualStream.read();
            int expected = expectedStream.read();
            assertThat("index " + index, actual, equalTo(expected));
            index++;
            if (actual < 0) {
                break;
            }
        }
    }

    static Iterator<StreamingPart> getMultipartFormParts(String boundary, byte[] multipartFormContents) throws IOException {
        return getMultipartFormParts(boundary.getBytes(StandardCharsets.UTF_8), multipartFormContents, StandardCharsets.UTF_8);
    }

    static Iterator<StreamingPart> getMultipartFormParts(byte[] boundary, byte[] multipartFormContents, Charset encoding) throws IOException {
        InputStream multipartFormContentsStream = new ByteArrayInputStream(multipartFormContents);
        return StreamingMultipartFormParts.Companion.parse(boundary, multipartFormContentsStream, encoding).iterator();
    }

    static StreamingPart assertFilePart(Iterator<StreamingPart> form, String fieldName, String fileName, String contentType, String contents) throws IOException {
        return assertFilePart(form, fieldName, fileName, contentType, contents, StandardCharsets.UTF_8);
    }

    static StreamingPart assertFilePart(Iterator<StreamingPart> form, String fieldName, String fileName, String contentType, String contents, Charset encoding) throws IOException {
        assertThereAreMoreParts(form);
        StreamingPart file = form.next();
        assertThat("file name", file.getFileName(), equalTo(fileName));
        assertThat("content type", file.getContentType(), equalTo(contentType));
        assertPartIsNotField(file);
        assertPart(fieldName, contents, file, encoding);
        return file;
    }

    static StreamingPart assertFieldPart(Iterator<StreamingPart> form, String fieldName, String fieldValue) throws IOException {
        return assertFieldPart(form, fieldName, fieldValue, StandardCharsets.UTF_8);
    }

    static StreamingPart assertFieldPart(Iterator<StreamingPart> form, String fieldName, String fieldValue, Charset encoding) throws IOException {
        assertThereAreMoreParts(form);
        StreamingPart field = form.next();
        assertPartIsFormField(field);
        assertPart(fieldName, fieldValue, field, encoding);
        return field;
    }

    static void assertPart(String fieldName, String fieldValue, StreamingPart StreamingPart, Charset encoding) throws IOException {
        assertThat("field name", StreamingPart.getFieldName(), equalTo(fieldName));
        assertThat("contents", StreamingPart.getContentsAsString(encoding, 4096), equalTo(fieldValue));
    }

    static void assertThereAreNoMoreParts(Iterator<StreamingPart> form) {
        assertFalse("Too many parts", form.hasNext());
    }

    static void assertThereAreMoreParts(Iterator<StreamingPart> form) {
        assertTrue("Not enough parts", form.hasNext());
    }

    static void assertPartIsFormField(StreamingPart field) {
        assertTrue("the StreamingPart is a form field", field.isFormField());
    }

    static void assertPartIsNotField(StreamingPart file) {
        assertFalse("the StreamingPart is not a form field", file.isFormField());
    }
}

File: http4k-multipart/src/test/kotlin/org/http4k/multipart/StreamingMultipartFormHappyTests.kt
package org.http4k.multipart

import com.natpryce.hamkrest.assertion.assertThat
import com.natpryce.hamkrest.equalTo
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.fail
import java.io.FileInputStream
import java.io.InputStream
import java.nio.charset.Charset
import java.nio.charset.StandardCharsets

class StreamingMultipartFormHappyTests {

    @Test
    fun uploadEmptyContents() {
        val boundary = "-----1234"
        val form = getMultipartFormParts(boundary, MultipartFormBuilder(boundary).stream())

        assertThereAreNoMoreParts(form)
    }

    @Test
    fun uploadEmptyFile() {
        val boundary = "-----2345"
        val form = getMultipartFormParts(boundary, MultipartFormBuilder(boundary)
            .file("aFile", "", "doesnt/matter", "".byteInputStream(), emptyList()).stream())

        assertFilePart(form, "aFile", "", "doesnt/matter", "")

        assertThereAreNoMoreParts(form)
    }

    @Test
    fun hasNextIsIdempotent() {
        val boundary = "-----2345"
        val form = getMultipartFormParts(boundary, MultipartFormBuilder(boundary)
            .file("aFile", "", "application/octet-stream", "".byteInputStream(), emptyList())
            .file("anotherFile", "", "application/octet-stream", "".byteInputStream(), emptyList()).stream())

        assertThereAreMoreParts(form)
        assertThereAreMoreParts(form)

        form.next()

        assertThereAreMoreParts(form)
        assertThereAreMoreParts(form)

        form.next()

        assertThereAreNoMoreParts(form)
        assertThereAreNoMoreParts(form)
    }

    @Test
    fun uploadEmptyField() {
        val boundary = "-----3456"
        val form = getMultipartFormParts(boundary, MultipartFormBuilder(boundary)
            .field("aField", "", emptyList()).stream())

        assertFieldPart(form, "aField", "")

        assertThereAreNoMoreParts(form)
    }

    @Test
    fun uploadSmallFile() {
        val boundary = "-----2345"
        val form = getMultipartFormParts(boundary, MultipartFormBuilder(boundary)
            .file("aFile", "file.name", "application/octet-stream", "File contents here".byteInputStream(), emptyList()).stream())

        assertFilePart(form, "aFile", "file.name", "application/octet-stream", "File contents here")

        assertThereAreNoMoreParts(form)
    }

    @Test
    fun uploadFileWithLowercaseContentDisposition() {
        val boundary = "-----2345"
        val form = getMultipartFormParts(boundary, MultipartFormBuilder(boundary)
            .part("File contents here".byteInputStream(),
                listOf(
                    "content-disposition" to "form-data; name=\"aFile\"; filename=\"file.name\"",
                    "Content-Type" to "application/octet-stream")
            )
            .stream())

        assertFilePart(form, "aFile", "file.name", "application/octet-stream", "File contents here")

        assertThereAreNoMoreParts(form)
    }

    @Test
    fun uploadSmallFileAsAttachment() {
        val boundary = "-----4567"
        val form = getMultipartFormParts(boundary, MultipartFormBuilder(boundary)
            .file("beforeFile", "before.txt", "application/json", "[]".byteInputStream(), emptyList())
            .startMultipart("multipartFieldName", "7890")
            .attachment("during.txt", "plain/text", "Attachment contents here", emptyList())
            .attachment("during2.txt", "plain/text", "More text here", emptyList())
            .endMultipart()
            .file("afterFile", "after.txt", "application/json", "[]".byteInputStream(), emptyList())
            .stream())

        assertFilePart(form, "beforeFile", "before.txt", "application/json", "[]")
        assertFilePart(form, "multipartFieldName", "during.txt", "plain/text", "Attachment contents here")
        assertFilePart(form, "multipartFieldName", "during2.txt", "plain/text", "More text here")
        assertFilePart(form, "afterFile", "after.txt", "application/json", "[]")

        assertThereAreNoMoreParts(form)
    }

    @Test
    fun uploadSmallField() {
        val boundary = "-----3456"
        val form = getMultipartFormParts(boundary, MultipartFormBuilder(boundary)
            .field("aField", "Here is the value of the field\n", emptyList()).stream())

        assertFieldPart(form, "aField", "Here is the value of the field\n")

        assertThereAreNoMoreParts(form)
    }

    @Test
    fun uploadMultipleFilesAndFields() {
        val boundary = "-----1234"
        val form = getMultipartFormParts(boundary,
            MultipartFormBuilder(boundary)
                .file("file", "foo.tab", "text/whatever", "This is the content of the file\n".byteInputStream(), emptyList())
                .field("field", "fieldValue" + CR_LF + "with cr lf", emptyList())
                .field("multi", "value1", emptyList())
                .file("anotherFile", "BAR.tab", "text/something", "This is another file\n".byteInputStream(), emptyList())
                .field("multi", "value2", emptyList())
                .stream())

        assertFilePart(form, "file", "foo.tab", "text/whatever", "This is the content of the file\n")
        assertFieldPart(form, "field", "fieldValue" + CR_LF + "with cr lf")
        assertFieldPart(form, "multi", "value1")
        assertFilePart(form, "anotherFile", "BAR.tab", "text/something", "This is another file\n")
        assertFieldPart(form, "multi", "value2")

        assertThereAreNoMoreParts(form)
    }

    @Test
    fun uploadFieldsWithMultilineHeaders() {
        val boundary = "-----1234"
        val form = getMultipartFormParts(boundary,
            MultipartFormBuilder(boundary)
                .part("Content-Disposition: form-data; \r\n" +
                    "\tname=\"field\"\r\n" +
                    "\r\n" +
                    "fieldValue", emptyList())
                .part("Content-Disposition: form-data;\r\n" +
                    "     name=\"multi\"\r\n" +
                    "\r\n" +
                    "value1", emptyList())
                .field("multi", "value2", emptyList())
                .stream())

        assertFieldPart(form, "field", "fieldValue")
        assertFieldPart(form, "multi", "value1")
        assertFieldPart(form, "multi", "value2")

        assertThereAreNoMoreParts(form)
    }

    @Test
    fun partsCanHaveLotsOfHeaders() {
        val boundary = "-----1234"
        val form = getMultipartFormParts(boundary,
            MultipartFormBuilder(boundary)
                .part("This is the content of the file\n",
                    listOf("Content-Disposition" to "form-data; name=\"fileFieldName\"; filename=\"filename.txt\"",
                        "Content-Type" to "plain/text",
                        "Some-header" to "some value"))
                .part("This is the content of the field\n",
                    listOf("Content-Disposition" to "form-data; name=\"fieldFieldName\"",
                        "Another-header" to "some-key=\"some-value\""))
                .stream())

        val file = assertFilePart(form, "fileFieldName", "filename.txt", "plain/text", "This is the content of the file\n")

        val fileHeaders = file.headers
        assertThat(fileHeaders.size, equalTo(3))
        assertThat(fileHeaders["Content-Disposition"], equalTo("form-data; name=\"fileFieldName\"; filename=\"filename.txt\""))
        assertThat(fileHeaders["Content-Type"], equalTo("plain/text"))
        assertThat(fileHeaders["Some-header"], equalTo("some value"))

        val field = assertFieldPart(form, "fieldFieldName", "This is the content of the field\n")

        val fieldHeaders = field.headers
        assertThat(fieldHeaders.size, equalTo(2))
        assertThat(fieldHeaders["Content-Disposition"], equalTo("form-data; name=\"fieldFieldName\""))
        assertThat(fieldHeaders["Another-header"], equalTo("some-key=\"some-value\""))

        assertThereAreNoMoreParts(form)
    }

    @Test
    fun closedPartsCannotBeReadFrom() {
        val boundary = "-----2345"
        val form = getMultipartFormParts(boundary, MultipartFormBuilder(boundary)
            .file("aFile", "file.name", "application/octet-stream", "File contents here".byteInputStream(), emptyList()).stream())

        val file = form.next()

        while (file.inputStream.read() > 0) {
            // keep reading.
        }

        assertThat(file.inputStream.read(), equalTo(-1))
        file.inputStream.close()
        file.inputStream.close() // can close multiple times
        try {
            val ignored = file.inputStream.read()
            fail("Should have complained that the StreamingPart has been closed " + ignored)
        } catch (e: AlreadyClosedException) {
            // pass
        }
    }

    @Test
    fun readingPartsContentsAsStringClosesStream() {
        val boundary = "-----2345"
        val form = getMultipartFormParts(boundary, MultipartFormBuilder(boundary)
            .file("aFile", "file.name", "application/octet-stream", "File contents here".byteInputStream(), emptyList()).stream())

        val file = form.next()
        file.contentsAsString

        try {
            val ignored = file.inputStream.read()
            fail("Should have complained that the StreamingPart has been closed " + ignored)
        } catch (e: AlreadyClosedException) {
            // pass
        }

        file.inputStream.close() // can close multiple times
    }

    @Test
    fun gettingNextPartClosesOldPart() {
        val boundary = "-----2345"
        val form = getMultipartFormParts(boundary, MultipartFormBuilder(boundary)
            .file("aFile", "file.name", "application/octet-stream", "File contents here".byteInputStream(), emptyList())
            .file("anotherFile", "your.name", "application/octet-stream", "Different file contents here".byteInputStream(), emptyList()).stream())

        val file1 = form.next()

        val file2 = form.next()

        assertThat(file1, !equalTo(file2))

        try {
            val ignored = file1.inputStream.read()
            fail("Should have complained that the StreamingPart has been closed " + ignored)
        } catch (e: AlreadyClosedException) {
            // pass
        }

        file1.inputStream.close() // can close multiple times

        assertThat(file2.contentsAsString, equalTo("Different file contents here"))
    }

    @Test
    fun canLoadComplexRealLifeSafariExample() {
        val parts = StreamingMultipartFormParts.parse(
            "----WebKitFormBoundary6LmirFeqsyCQRtbj".toByteArray(StandardCharsets.UTF_8),
            FileInputStream("examples/safari-example.multipart"),
            StandardCharsets.UTF_8
        ).iterator()

        assertFieldPart(parts, "articleType", "obituary")

        assertRealLifeFile(parts, "simple7bit.txt", "text/plain")
        assertRealLifeFile(parts, "starbucks.jpeg", "image/jpeg")
        assertRealLifeFile(parts, "utf8\uD83D\uDCA9.file", "application/octet-stream")
        assertRealLifeFile(parts, "utf8\uD83D\uDCA9.txt", "text/plain")
    }

    @Test
    fun canLoadComplexRealLifeChromeExample() {
        val parts = StreamingMultipartFormParts.parse(
            "----WebKitFormBoundaryft3FGhOMTYoOkCCc".toByteArray(StandardCharsets.UTF_8),
            FileInputStream("examples/chrome-example.multipart"),
            StandardCharsets.UTF_8
        ).iterator()

        assertFieldPart(parts, "articleType", "obituary")

        assertRealLifeFile(parts, "simple7bit.txt", "text/plain")
        assertRealLifeFile(parts, "starbucks.jpeg", "image/jpeg")
        assertRealLifeFile(parts, "utf8\uD83D\uDCA9.file", "application/octet-stream")
        assertRealLifeFile(parts, "utf8\uD83D\uDCA9.txt", "text/plain")
    }
}

val CR_LF = "\r\n"

internal fun assertRealLifeFile(parts: Iterator<StreamingPart>, fileName: String, contentType: String) {
    val file = parts.next()
    assertThat("field name", file.fieldName, equalTo("uploadManuscript"))
    assertThat("file name", file.fileName, equalTo(fileName))
    assertThat("content type", file.contentType, equalTo(contentType))
    assertPartIsNotField(file)
    compareStreamToFile(file)
}

internal fun compareStreamToFile(file: StreamingPart) {
    val formFile = file.inputStream
    compareStreamToFile(formFile, file.fileName)
}

fun compareStreamToFile(actualSream: InputStream, fileName: String?) {
    val original = FileInputStream("examples/" + fileName!!)
    compareOneStreamToAnother(actualSream, original)
}

fun compareOneStreamToAnother(actualStream: InputStream, expectedStream: InputStream) {
    var index = 0
    while (true) {
        val actual = actualStream.read()
        val expected = expectedStream.read()
        assertThat("index " + index, actual, equalTo(expected))
        index++
        if (actual < 0) {
            break
        }
    }
}

internal fun getMultipartFormParts(boundary: String, multipartFormContents: InputStream): Iterator<StreamingPart> =
    getMultipartFormParts(boundary.toByteArray(StandardCharsets.UTF_8), multipartFormContents, StandardCharsets.UTF_8)

internal fun getMultipartFormParts(boundary: ByteArray, multipartFormContents: InputStream, encoding: Charset): Iterator<StreamingPart> =
    StreamingMultipartFormParts.parse(boundary, multipartFormContents, encoding).iterator()

internal fun assertFilePart(form: Iterator<StreamingPart>, fieldName: String, fileName: String, contentType: String, contents: String, encoding: Charset = StandardCharsets.UTF_8): StreamingPart {
    assertThereAreMoreParts(form)
    val file = form.next()
    assertThat("file name", file.fileName, equalTo(fileName))
    assertThat("content type", file.contentType, equalTo(contentType))
    assertPartIsNotField(file)
    assertPart(fieldName, contents, file, encoding)
    return file
}

internal fun assertFieldPart(form: Iterator<StreamingPart>, fieldName: String, fieldValue: String, encoding: Charset = StandardCharsets.UTF_8): StreamingPart {
    assertThereAreMoreParts(form)
    val field = form.next()
    assertPartIsFormField(field)
    assertPart(fieldName, fieldValue, field, encoding)
    return field
}

internal fun assertPart(fieldName: String, fieldValue: String, StreamingPart: StreamingPart, encoding: Charset) {
    assertThat("field name", StreamingPart.fieldName, equalTo(fieldName))
    assertThat("contents", StreamingPart.inputStream.reader(encoding).use { it.readText() }, equalTo(fieldValue))
}

internal fun assertThereAreNoMoreParts(form: Iterator<StreamingPart>) {
    assertThat("Too many parts", form.hasNext(), equalTo(false))
}

internal fun assertThereAreMoreParts(form: Iterator<StreamingPart>) {
    assertThat("Not enough parts", form.hasNext(), equalTo(true))
}

internal fun assertPartIsFormField(field: StreamingPart) {
    assertThat("the StreamingPart is a form field", field.isFormField, equalTo(true))
}

internal fun assertPartIsNotField(file: StreamingPart) {
    assertThat("the StreamingPart is not a form field", file.isFormField, equalTo(false))
}
